File: c_glib/arrow-glib/struct-data-type.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <arrow-glib/data-type.hpp>
#include <arrow-glib/field.hpp>
#include <arrow-glib/struct-data-type.h>

G_BEGIN_DECLS

/**
 * SECTION: struct-data-type
 * @short_description: Struct data type
 *
 * #GArrowStructDataType is a class for struct data type.
 */

G_DEFINE_TYPE(GArrowStructDataType,                \
              garrow_struct_data_type,             \
              GARROW_TYPE_DATA_TYPE)

static void
garrow_struct_data_type_init(GArrowStructDataType *object)
{
}

static void
garrow_struct_data_type_class_init(GArrowStructDataTypeClass *klass)
{
}

/**
 * garrow_struct_data_type_new:
 * @fields: (element-type GArrowField): The fields of the struct.
 *
 * Returns: The newly created struct data type.
 */
GArrowStructDataType *
garrow_struct_data_type_new(GList *fields)
{
  std::vector<std::shared_ptr<arrow::Field>> arrow_fields;
  for (GList *node = fields; node; node = g_list_next(node)) {
    auto field = GARROW_FIELD(node->data);
    auto arrow_field = garrow_field_get_raw(field);
    arrow_fields.push_back(arrow_field);
  }

  auto arrow_data_type = std::make_shared<arrow::StructType>(arrow_fields);
  GArrowStructDataType *data_type =
    GARROW_STRUCT_DATA_TYPE(g_object_new(GARROW_TYPE_STRUCT_DATA_TYPE,
                                         "data-type", &arrow_data_type,
                                         NULL));
  return data_type;
}

G_END_DECLS

File: ruby/red-arrow/lib/arrow/struct-data-type.rb
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

require "arrow/field-containable"

module Arrow
  class StructDataType
    include FieldContainable

    alias_method :initialize_raw, :initialize
    private :initialize_raw

    # Creates a new {Arrow::StructDataType}.
    #
    # @overload initialize(fields)
    #
    #   @param fields [::Array<Arrow::Field, Hash>] The fields of the
    #     struct data type. You can also specify field description as
    #     a field. You can mix {Arrow::Field} and field description.
    #
    #     See {Arrow::Field.new} how to specify field description.
    #
    #   @example Create a struct data type with {Arrow::Field}s
    #     visible_field = Arrow::Field.new("visible", :boolean)
    #     count_field = Arrow::Field.new("count", :int32)
    #     Arrow::StructDataType.new([visible_field, count_field])
    #
    #   @example Create a struct data type with field descriptions
    #     field_descriptions = [
    #       {name: "visible", type: :boolean},
    #       {name: "count", type: :int32},
    #     ]
    #     Arrow::StructDataType.new(field_descriptions)
    #
    #   @example Create a struct data type with {Arrow::Field} and field description
    #     fields = [
    #       Arrow::Field.new("visible", :boolean),
    #       {name: "count", type: :int32},
    #     ]
    #     Arrow::StructDataType.new(fields)
    #
    # @overload initialize(fields)
    #
    #   @param fields [Hash{String, Symbol => Arrow::DataType, Hash}]
    #     The pairs of field name and field data type of the struct
    #     data type. You can also specify data type description by
    #     `Hash`. You can mix {Arrow::DataType} and data type description.
    #
    #     See {Arrow::DataType.resolve} how to specify data type
    #     description.
    #
    #   @example Create a struct data type with {Arrow::DataType}s
    #     fields = {
    #       "visible" => Arrow::BooleanDataType.new,
    #       "count" => Arrow::Int32DataType.new,
    #     }
    #     Arrow::StructDataType.new(fields)
    #
    #   @example Create a struct data type with data type descriptions
    #     fields = {
    #       "visible" => :boolean,
    #       "count" => {type: :int32},
    #     }
    #     Arrow::StructDataType.new(fields)
    #
    #   @example Create a struct data type with {Arrow::DataType} and data type description
    #     fields = {
    #       "visible" => Arrow::BooleanDataType.new,
    #       "count" => {type: :int32},
    #     }
    #     Arrow::StructDataType.new(fields)
    #
    # @overload initialize(description)
    #
    #   @param description [Hash] The description of the struct data
    #     type. It must have `:fields` value.
    #
    #   @option description
    #     [::Array<Arrow::Field, Hash>,
    #      Hash{String, Symbol => Arrow::DataType, Hash, String, Symbol}]
    #     :fields The fields of the struct data type.
    #
    #   @example Create a struct data type with {Arrow::Field} and field description
    #     fields = [
    #       Arrow::Field.new("visible", :boolean),
    #       {name: "count", type: :int32},
    #     ]
    #     Arrow::StructDataType.new(fields: fields)
    #
    #   @example Create a struct data type with {Arrow::DataType} and data type description
    #     fields = {
    #       "visible" => Arrow::BooleanDataType.new,
    #       "count" => {type: :int32},
    #     }
    #     Arrow::StructDataType.new(fields: fields)
    def initialize(fields)
      if fields.is_a?(Hash) and fields.key?(:fields)
        description = fields
        fields = description[:fields]
      end
      if fields.is_a?(Hash)
        fields = fields.collect do |name, data_type|
          Field.new(name, data_type)
        end
      else
        fields = fields.collect do |field|
          field = Field.new(field) unless field.is_a?(Field)
          field
        end
      end
      initialize_raw(fields)
    end

    alias_method :[], :find_field
  end
end
